rollup-interface/src/state_machine/da.rs

```diff
@@ -56,14 +56,68 @@ pub trait DaVerifier {
     ) -> Result<(), Self::Error>;
 }
 
+pub struct BlobDataAndHash<B: Buf> {
+    pub data: BufWithCounter<B>,
+    pub hash: [u8; 32],
+}
+
+// Simple structure that implements a buffer that counts the number of bytes read from the beginning
+// Useful for the partial blob reading optimization: we know for each blob how many bytes have been read from the beginning
+#[derive(Debug, Clone, Serialize, Deserialize, BorshDeserialize, BorshSerialize, PartialEq)]
+pub struct BufWithCounter<B: Buf> {
+    inner: B,
+    counter: usize,
+}
+
+impl<B: Buf> BufWithCounter<B> {
+    pub fn new(inner: B) -> Self {
+        BufWithCounter { inner, counter: 0 }
+    }
+
+    pub fn inner(&self) -> &B {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut B {
+        &mut self.inner
+    }
+
+    pub fn counter(&self) -> usize {
+        self.counter
+    }
+}
+
+impl<B: Buf> Buf for BufWithCounter<B> {
+    fn remaining(&self) -> usize {
+        self.inner.remaining()
+    }
+
+    fn chunk(&self) -> &[u8] {
+        self.inner.chunk()
```

Comment:
This function looks like it could cause a soundness problem - using it, an STF could read data from the buffer without advancing the counter. If the prover was malicious, they could tamper with that data and since the counter didn't advance it might not be checked by the DA verifier.

I see a couple of options to fix it:
- add an extra counter to let us track the furthest byte read in addition to the location of the cursor
- Modify the API of the STF to accept `impl Read` instead of `impl Buf` (this will cause a slight efficiency loss)
- (not recommended) Try to guarantee that a `chunk` always corresponds to a share of the underlying DA layer, so that any data returned by `chunk` will automatically get authenticated

Reply:
Fixed by the last commit, I went with the second approach as the extra counter wasn't feasible (the chunk method doesn't take a mutable version of self)